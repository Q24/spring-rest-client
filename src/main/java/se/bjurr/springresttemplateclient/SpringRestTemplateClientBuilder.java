package se.bjurr.springresttemplateclient;

import java.lang.reflect.Proxy;
import org.springframework.http.HttpHeaders;
import org.springframework.web.client.RestTemplate;
import se.bjurr.springresttemplateclient.handler.SpringRestTemplateClientInvocationHandler;

public final class SpringRestTemplateClientBuilder {
  private SpringRestTemplateClientBuilder() {}

  public static class Builder<T> {
    private final Class<T> api;
    private RestTemplate restTemplate = new RestTemplate();
    private HttpHeaders headers = new HttpHeaders();
    private String url;

    public Builder(final Class<T> api) {
      if (api == null) {
        throw new IllegalArgumentException("api cannot be null");
      }
      this.api = api;
    }

    public Builder<T> setRestTemplate(final RestTemplate restTemplate) {
      this.restTemplate = restTemplate;
      return this;
    }

    public Builder<T> setUrl(final String url) {
      this.url = url;
      return this;
    }

    public Builder<T> setHeaders(final HttpHeaders headers) {
      this.headers = headers;
      return this;
    }

    public Builder<T> setHeader(final String headerName, final String headerValue) {
      this.headers.add(headerName, headerValue);
      return this;
    }

    @SuppressWarnings("unchecked")
    public T build() {
      if (this.url == null) {
        throw new IllegalArgumentException("url cannot be null");
      }
      final SpringRestTemplateClientInvocationHandler<T> invocationHandler =
          new SpringRestTemplateClientInvocationHandler<>(
              this.url, this.restTemplate, this.headers);

      return (T)
          Proxy.newProxyInstance(
              this.api.getClassLoader(), new Class[] {this.api}, invocationHandler);
    }
  }

  public static <T> Builder<T> springRestTemplateClientBuilder(final Class<T> clazz) {
    return new Builder<>(clazz);
  }

  public static <T> Builder<T> create(final Class<T> clazz) {
    return springRestTemplateClientBuilder(clazz);
  }
}
