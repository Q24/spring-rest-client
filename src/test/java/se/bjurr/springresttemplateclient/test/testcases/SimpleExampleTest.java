package se.bjurr.springresttemplateclient.test.testcases;

import java.util.Arrays;
import org.junit.Test;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.web.client.RestTemplate;
import se.bjurr.springresttemplateclient.SpringRestTemplateClientBuilder;
import se.bjurr.springresttemplateclient.test.spec.api.StoreApi;
import se.bjurr.springresttemplateclient.test.utils.BaseApiTest;

public class SimpleExampleTest extends BaseApiTest<StoreApi> {

  @Override
  public Class<StoreApi> getSutClass() {
    return StoreApi.class;
  }

  @Test
  public void testExample() {
    final RestTemplate restTemplate = new RestTemplate();
    restTemplate.setMessageConverters(Arrays.asList(new MappingJackson2HttpMessageConverter()));

    final StoreApi storeApi =
        SpringRestTemplateClientBuilder.create(StoreApi.class)
            .setRestTemplate(restTemplate)
            .setUrl(this.getMockUrl())
            .build();

    storeApi.deleteOrder(1234L);

    this.verify();
  }
}
