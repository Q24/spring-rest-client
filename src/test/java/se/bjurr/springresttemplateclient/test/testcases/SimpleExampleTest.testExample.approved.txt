[ {
  "url" : "http://localhost:X/store/order/1234",
  "method" : "DELETE",
  "headers" : {
    "Accept" : "application/json, application/*+json",
    "Connection" : "keep-alive"
  },
  "cookies" : { },
  "body" : ""
} ]